### PR TITLE
Pr91x l1t prescales from payloads rebased

### PR DESCRIPTION
**Pull request by rekovic**
Rebased #17721.  
> Updated L1TGlobalUtil to now also access L1T prescales from payloads, using GlobalTag.

### CODE REVIEW

**Comment by rekovic** [comment]
please test

**Comment by rekovic** [comment]
+1

**Comment by perrotta** [comment]
+1
- Reco enters through PhysicsTools/PatAlgos/python/triggerLayer1/triggerProducer_cfi.py , where one parameter was added to a configuration for a PATTriggerProducer
- Checked in jenkins: no changes
- Comments received from L1/HLT are now integrated
(repeating here the same comment as in https://github.com/cms-sw/cmssw/pull/17721#issuecomment-295248696; further to that, I'd add the observation that there are a lot of thread-unsafe cout's in the L1 part of this PR)

**Comment by rekovic** [comment]
@perrotta Thanks for the review and the signature. 

> ... further to that, I'd add the observation that there are a lot of thread-unsafe cout's in the L1 part of this PR)

Here, `cout` is used only with a boolean switch `dumpTriggerResults_`, which is set to `false` by default. This should be OK.

**Comment by dmitrijus** [comment]
+1

File: L1Trigger/L1TGlobal/src/L1TGlobalUtil.cc [hunk 1]
```diff
@@ -97,84 +99,99 @@ void l1t::L1TGlobalUtil::retrieveL1Setup(const edm::EventSetup& evSetup) {
...
+      // Get the algorithm name
+      std::string algName = itAlgo->first;
+      int algBit = (itAlgo->second).getIndex(); //algoBitNumber();
+
+      (m_prescales[algBit]).first  = algName;
+      (m_prescales[algBit]).second = prescaleSet[algBit];
+
+      LogDebug("l1t|Global")<< "Number of bunch crossings stored: " << (*m_triggerMaskAlgoTrig).size() << endl;
+
+      std::map<int, std::vector<int> > triggerAlgoMaskAlgoTrig = *m_triggerMaskAlgoTrig;
```
Comment (davidlange6) [inline L1Trigger/L1TGlobal/src/L1TGlobalUtil.cc]: Hi @rekovic - please avoid this copy

File: L1Trigger/L1TGlobal/interface/L1TGlobalUtil.h [hunk 2]
```diff
@@ -179,35 +175,36 @@ class L1TGlobalUtil {
...
     std::vector<std::vector<int> > m_initialPrescaleFactorsAlgoTrig;
     const std::vector<std::vector<int> >* m_prescaleFactorsAlgoTrig;
-    std::vector<unsigned int>  m_initialTriggerMaskAlgoTrig;
-    const std::vector<unsigned int>*  m_triggerMaskAlgoTrig;
-    std::vector<int>   m_initialTriggerMaskVetoAlgoTrig;
-    const std::vector<int>*  m_triggerMaskVetoAlgoTrig;
-    
-    // access to the results block from uGT 
+    const std::map<int, std::vector<int> > m_initialTriggerMaskAlgoTrig;
+    const std::map<int, std::vector<int> >*  m_triggerMaskAlgoTrig; // vector stores the BX
```
Comment (davidlange6) [inline L1Trigger/L1TGlobal/interface/L1TGlobalUtil.h]: why a pointer? (actually all of these being pointers seems to cause considerable problems in code clarify below)

File: L1Trigger/L1TGlobal/src/L1TGlobalUtil.cc [hunk 3]
```diff
@@ -191,56 +208,90 @@ void l1t::L1TGlobalUtil::retrieveL1Event(const edm::Event& iEvent, const edm::Ev
...
 	   bool decisionFinal     = algBlk->getAlgoDecisionFinal(algBit);
 	   (m_decisionsFinal[algBit]).first  = algName;
 	   (m_decisionsFinal[algBit]).second = decisionFinal;
+
+	   (m_prescales[algBit]).first  = algName;
+	   (m_prescales[algBit]).second = prescaleSet[algBit];
+
+	   LogDebug("l1t|Global") << "Number of bunch crossings stored: " <<  (*m_triggerMaskAlgoTrig).size() << endl;
+
+	   std::map<int, std::vector<int> > triggerAlgoMaskAlgoTrig = *m_triggerMaskAlgoTrig;
```
Comment (davidlange6) [inline L1Trigger/L1TGlobal/src/L1TGlobalUtil.cc]: and avoid this copy (every event?)

File: L1Trigger/L1TGlobal/src/L1TGlobalUtil.cc [hunk 4]
```diff
@@ -97,84 +99,99 @@ void l1t::L1TGlobalUtil::retrieveL1Setup(const edm::EventSetup& evSetup) {
...
-	  (m_masks[algBit]).second = (*m_triggerMaskAlgoTrig)[algBit];	  
+	m_prescaleFactorsAlgoTrig = &(m_l1GtPrescalesVetoes->prescaleTable());
+	m_numberPhysTriggers = (*m_prescaleFactorsAlgoTrig)[0].size(); // assumes all prescale columns are the same length
 
-	  (m_vetoMasks[algBit]).first  = algName;
-	  (m_vetoMasks[algBit]).second = (*m_triggerMaskVetoAlgoTrig)[algBit];
-       }
-       
-      m_filledPrescales = true;
+	m_triggerMaskAlgoTrig   = &(m_l1GtPrescalesVetoes->triggerAlgoBxMask());
```
Comment (davidlange6) [inline L1Trigger/L1TGlobal/src/L1TGlobalUtil.cc]: so triggerAlgoBxMask() is already returning the right thing - a const ref. why it not be used directly?

**Comment by davidlange6** [comment]
-1
pending answers to comments

File: L1Trigger/L1TGlobal/src/L1TGlobalUtil.cc (hunk 1 above)
Comment (rekovic) [inline L1Trigger/L1TGlobal/src/L1TGlobalUtil.cc]: Will change in next commit.

File: L1Trigger/L1TGlobal/src/L1TGlobalUtil.cc (hunk 3 above)
Comment (rekovic) [inline L1Trigger/L1TGlobal/src/L1TGlobalUtil.cc]: Will change in next commit.

File: L1Trigger/L1TGlobal/interface/L1TGlobalUtil.h (hunk 2 above)
Comment (rekovic) [inline L1Trigger/L1TGlobal/interface/L1TGlobalUtil.h]: Used to point to either a const ref. returned by triggerAlgoBxMask() or the Masks returned from file (dummy for the moment).

File: L1Trigger/L1TGlobal/src/L1TGlobalUtil.cc (hunk 4 above)
Comment (rekovic) [inline L1Trigger/L1TGlobal/src/L1TGlobalUtil.cc]: See https://github.com/cms-sw/cmssw/pull/18428#discussion_r112989269

**Comment by rekovic** [comment]
please test

**Comment by rekovic** [comment]
+1

**Comment by perrotta** [comment]
+1
Changes applied in the last commits are fine with reco 
I can confirm my "+1" as in  https://github.com/cms-sw/cmssw/pull/18428#issuecomment-296124937

**Comment by rekovic** [comment]
@dmitrijus  please re-sign

**Comment by dmitrijus** [comment]
+1

**Pull request merged**